<?xml version="1.0" encoding="UTF-8"?>
<root BTCPP_format="3"
      main_tree_to_execute="MainTree">
  <BehaviorTree ID="MainTree">
    <ReactiveSequence>
      <SequenceStar>
        <GetTagList tag_list="{tag_list}"/>
        <ForEachTag tag_list="{tag_list}"
                    current_tag="{current_tag}">
          <SequenceStar>
            <RetryUntilSuccessful num_attempts="3">
              <BoxTagOK tag_id="{current_tag}"
                        box_pos="{pos}"
                        box_quat="{quat}"/>
            </RetryUntilSuccessful>
            <ComputeTargetPose box_pos="{pos}"
                               box_quat="{quat}"
                               target_box_offset="{pre_grab_box_offset}"
                               target_pose="{target_pose_pre_grasp}"/>
            <MakePlan target_pose="{target_pose_pre_grasp}"
                      planed_trajectory="{planed_trajectory_grasp}"/>
            <MoveToDestination planed_trajectory="{planed_trajectory_grasp}"
                               move_type="0"/>
            <RetryUntilSuccessful num_attempts="3">
              <GraspBox box_pos="{pos}"
                        box_quat="{quat}"
                        box_size="0.3,0.3,0.3"
                        grasp_type="0"/>
            </RetryUntilSuccessful>
            <ComputeTargetPlacePose tag_id="{current_tag}"
                                    target_place_pose="{target_place_pose}"/>
            <MakePlan target_pose="{target_place_pose}"
                      planed_trajectory="{planed_trajectory_place}"/>
            <MoveToDestination planed_trajectory="{planed_trajectory_place}"
                               move_type="1"/>
          </SequenceStar>
        </ForEachTag>
      </SequenceStar>
    </ReactiveSequence>
  </BehaviorTree>

  <!-- Description of Node Models (used by Groot) -->
  <TreeNodesModel>
    <Condition ID="BoxTagOK"
               editable="true">
      <input_port name="tag_id"/>
      <output_port name="box_pos"/>
      <output_port name="box_quat"/>
    </Condition>
    <Action ID="ComputeTargetPlacePose"
            editable="true">
      <input_port name="tag_id"/>
      <output_port name="target_place_pose"/>
    </Action>
    <Condition ID="ComputeTargetPose"
               editable="true">
      <input_port name="box_pos"/>
      <input_port name="box_quat"/>
      <input_port name="target_box_offset"/>
      <output_port name="target_pose"/>
    </Condition>
    <Decorator ID="ForEachTag"
               editable="true">
      <input_port name="tag_list"/>
      <output_port name="current_tag">id,pos,quat</output_port>
    </Decorator>
    <Condition ID="GetTagList"
               editable="true">
      <output_port name="tag_list"/>
    </Condition>
    <Action ID="GraspBox"
            editable="true">
      <input_port name="box_pos"/>
      <input_port name="box_quat">xyzw</input_port>
      <input_port name="box_size"/>
      <input_port name="grasp_type">0:grasp, 1:put</input_port>
    </Action>
    <Action ID="MakePlan"
            editable="true">
      <input_port name="target_pose"/>
      <output_port name="planed_trajectory"/>
    </Action>
    <Action ID="MoveToDestination"
            editable="true">
      <input_port name="planed_trajectory"/>
      <input_port name="move_type"
                  default="0">int</input_port>
    </Action>
  </TreeNodesModel>

</root>
